compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp

```diff
@@ -244,107 +229,292 @@ static bool setPipeliningMarkers(scf::ForOp forOp, bool pipelineStoreStage) {
   return copyToWorkgroupMemory;
 }
 
-/// Return true if the op is used as operand `index` of an mma.sync op.
-static bool isMMAOperand(Operation* op, int64_t index) {
-  OpOperand* use = &(*((op->getUses()).begin()));
-  if (auto extract = dyn_cast<vector::ExtractStridedSliceOp>(use->getOwner())) {
-    use = &(*((extract->getUses()).begin()));
+/// Loads from Shared Memory and the MMA operations on registers for a kgroup.
+struct WarpMmaOp {
+  // Load matrixA from Shared Memory to registers.
+  llvm::SetVector<Operation*> loadOperationsA;
+  // Load matrixB from Shared Memory to registers.
+  llvm::SetVector<Operation*> loadOperationsB;
+  // Warp-level MMA operations on registers.
+  llvm::SetVector<Operation*> mmaOperations;
+};
+
+/// Structure to hold the matmul's mainloop information:
+/// Seperates the mmasync operations into kgroups and collects the Shared Memory
+/// loads for each kgroup. This information is used to pipeline the mainloop and
+/// to generate an optimal schedule interleaving Global Memory loads, Shared
+/// Memory loads, and math operations.
+struct MainLoopInfo {
+  //
+  // Type definitions
+  //
+  using KgroupToWarpMmaMap = std::unordered_map<int, WarpMmaOp>;
+
+  //
+  // Data members
+  //
+
+  // Mainloop operations GlobalMemory -> SharedMemory
+  SetVector<Operation*> copyGlobalToSharedOps;
+  SetVector<Operation*> asyncCreateGroupOp;
+  SetVector<Operation*> barrierOps;
+  SetVector<Operation*> asyncWaitOps;
+
+  // Warp-level MMA operations SharedMemory -> Registers -> MMA.
+  // Map: kgroupIdx -> WarpMmaOp
+  KgroupToWarpMmaMap warpOperations;
+
+  std::vector<llvm::SetVector<Operation*>> ldmatrixOpDeps;
+  std::vector<llvm::SetVector<Operation*>> mmasyncOpDeps;
+  llvm::SetVector<Operation*> copyGlobalToSharedOpDeps;
+
+  // Map: Operation -> kgroupIdx
+  std::unordered_map<Operation*, int> opToKgroupMap;
+
+  // Set to track the dependencies already seen to a backward slice.
+  llvm::SetVector<Operation*> seenDepOps;
+
+  // Set to track the mma operations in forward slice to count kgroups and
+  // populate the warp-level WarpMmaOp kgroupIdx -> WarpMmaOp map.
+  llvm::SetVector<Operation*> seenMmaOps;
```

Comment:
I think we should be able to simplify those data structure.

Do we really need the reverse mapping `opToKgroupMap`? The mmasync are already sorted by groups, can't you just loop through.
Also can we merge the operations with their dependency into one SetVector rather than having two?

Also in general you want to prefer `llvm::DenseMap` to `std::unordered_map` but I think hopefully we can remove those maps here anyway.